src/coreComponents/constitutive/solid/AuxiliaryFunctionsSpectral.hpp

```diff
@@ -0,0 +1,460 @@
+#include <iostream>
+#include "LvArray/src/output.hpp"
+#include "LvArray/src/tensorOps.hpp"
+namespace geosx
+{
+
+//get the positive part only of tensor T using spectral split
+GEOSX_HOST_DEVICE inline
+void PositivePartOfTensor( real64 (& eigs)[3], real64 (& eigvecs)[3][3], real64 (& positivePart)[6] )
+{
+  real64 positiveEigs[6]={};
+  for( int i=0; i < 3; i++ )
+  {
+    positiveEigs[i] = std::max( 0.0, eigs[i] );
+  }
+  LvArray::tensorOps::Rij_eq_AikSymBklAjl< 3 >( positivePart, eigvecs, positiveEigs );
+}
+
+GEOSX_HOST_DEVICE inline
+void NegativePartOfTensor( real64 (& eigs)[3], real64 (& eigvecs)[3][3], real64 (& negativePart)[6] )
+{
+  real64 negativeEigs[6]={};
+  for( int i=0; i < 3; i++ )
+  {
+    negativeEigs[i] = std::min( 0.0, eigs[i] );
+  }
+  LvArray::tensorOps::Rij_eq_AikSymBklAjl< 3 >( negativePart, eigvecs, negativeEigs );
+}
+
+GEOSX_HOST_DEVICE inline
+real64 doubleContraction( real64 (& A)[6], real64 (& B)[6] )
+{
+  real64 ans = 0;
+  for( int i=0; i < 6; i++ )
+  {
+    if( i < 3 )
+    {
+      ans = ans + A[i]*B[i];
+    }
+    else
+    {
+      ans = ans + 2*A[i]*B[i];
+    }
+  }
+  return ans;
+}
+
+GEOSX_HOST_DEVICE inline
+void recoverStrainFromStress( arraySlice1d< real64 > const stress, real64 (& strain)[6], real64 const K, real64 const mu )
+{
+  real64 E = 9*K*mu / (3*K + mu);
+  real64 nu = (3*K - 2*mu) / (6*K + 2*mu);
+  strain[0] = (stress[0] - nu*(stress[1] + stress[2]))/E;
+  strain[1] = (stress[1] - nu*(stress[0] + stress[2]))/E;
+  strain[2] = (stress[2] - nu*(stress[0] + stress[1]))/E;
+  strain[3] = (1 + nu)*stress[3]/E;
+  strain[4] = (1 + nu)*stress[4]/E;
+  strain[5] = (1 + nu)*stress[5]/E;
+}
+
+GEOSX_HOST_DEVICE inline
+int voigt( int voigtIndex, int pairIndex )
+{
+  if( pairIndex != 1 && pairIndex != 2 )
+  {
+    GEOSX_ERROR( "Index of the pair must be 1 or 2" );
+  }
+  switch( voigtIndex )
+  {
+    case 0:
+    {
+      if( pairIndex == 1 )
+      {
+        return 1;
+      }
+      else
+      {
+        return 1;
+      }
+    }
+    break;
+    case 1:
+    {
+      if( pairIndex == 1 )
+      {
+        return 2;
+      }
+      else
+      {
+        return 2;
+      }
+    }
+    break;
+    case 2:
+    {
+      if( pairIndex == 1 )
+      {
+        return 3;
+      }
+      else
+      {
+        return 3;
+      }
+    }
+    break;
+    case 3:
+    {
+      if( pairIndex == 1 )
+      {
+        return 2;
+      }
+      else
+      {
+        return 3;
+      }
+    }
+    break;
+    case 4:
+    {
+      if( pairIndex == 1 )
+      {
+        return 1;
+      }
+      else
+      {
+        return 3;
+      }
+    }
+    break;
+    case 5:
+    {
+      if( pairIndex == 1 )
+      {
+        return 1;
+      }
+      else
+      {
+        return 2;
+      }
+    }
+    break;
+    default:
+      GEOSX_ERROR( "Voigt index must be from 0 to 5" );
+      return 1000000;
+  }
+}
+
+GEOSX_HOST_DEVICE inline
+real64 heaviside( real64 x )
+{
+  if( std::abs( x ) < 1e-12 )
```

Comment:
This seems arbitrary. Can you specify a normalization?

Reply:
I am not sure if this possible, since there is only a single argument in the function. The idea is to normalize before calling the heaviside function.